### PR TITLE
feat: generate 2fa backup code

### PR DESCRIPTION
**Pull request by olamstevy**
## What does the PR do?

This PR is about generating a backup code for logged in users who may later not have access to their devices to obtain totp. The endpoint returns errors if:

- [x] User is not logged in.
- [x] 2 Factor authentication is not enabled.
- [x] provided totp is not valid or incorrect
- [x] provided password failed verification

### How should this be manually tested?

- Login to a user account, and enable two factor authentication

- Make a POST request to the `api/v1/auth/2fa/backup-codes` endpoint with the following payload:
`Authorization: Bearer <access_token>`
```
{
    "totp_code": "<totp_code on your device>",
    "password": "<password>"
}
```

#### Successful Response
```json
{
    "status_code": 200,
    "message": "New backup codes generated",
    "data": {
        "backup_codes": [code, code, code, code, code],
    }
}
```

#### Authentication Error:
When the password provided is not correct:
```json
{
    "status_code": 401,
    "error": "Unauthorized",
    "message": "Authentication required",
}
```

#### Invalid totp_code Error
```json
{
    "status_code": 400,
    "error": "Invalid Request",
    "message": "TOTP code provided is invalid",
}
```

#### When 2 factor authentication is not enabled
```
{
    "status_code": 400,
    "error": "Invalid Request",
    "message": "2-Factor Authentication has not been enabled"
}
```

### Task

- [x] Create a POST endpoint /api/v1/auth/2fa/backup-codes to generate backup code for logged in users.
- [x] Implement server-side logic to handle successful and failed post request.
- [x] Ensure the user has the proper authentication to generate the codes.
- [x] Update the essential details such as backup codes and totp_code in the database.
- [x] Write tests for generating 2fa backup codes.
- [x] Update API documentation to reflect the new endpoint and its usage.

### CODE REVIEW

File: src/modules/auth/auth.controller.ts
Comment (Heba-WebDev) [inline src/modules/auth/auth.controller.ts:1]: You should return right away without using response.status.send, as nestjs detects when the handler is using @ Response

Comment (Heba-WebDev) [reply src/modules/auth/auth.controller.ts:1]: Also remove the Request, no need for it. Please read the Nestjs docs

**Comment by Heba-WebDev** [comment]
Left you comments and you're also failing the CI/CD

File: src/modules/auth/auth.controller.ts
Comment (olamstevy) [reply src/modules/auth/auth.controller.ts:1]: Thanks. Working on it

Comment (olamstevy) [reply src/modules/auth/auth.controller.ts:1]: I left the request to get the user id

**Review by Heba-WebDev — DISMISSED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Comment by olamstevy** [comment]
Feature has been implemented.

**Pull request closed without merging**